src/number_theory/diophantine_approximation.lean

```diff
@@ -265,3 +284,317 @@ begin
   norm_cast,
 end
+
+/-!
+### Legendre's Theorem on Rational Approximation
+
+We prove **Legendre's Theorem** on rational approximation: If $\xi$ is a real number and
+$x/y$ is a rational number such that $|\xi - x/y| < 1/(2y^2)$,
+then $x/y$ is a convergent of the continued fraction expansion of $\xi$.
+
+The proof is by induction. However, the induction proof does not work with the
+statement as given, since the assumption is too weak to imply the corresponding
+statement for the application of the induction hypothesis. This can be remedied
+by making the statement slightly stronger. Namely, we assume that $|\xi - x/y| < 1/(y(2y-1))$
+when $y \ge 2$ and $-\frac{1}{2} < \xi - x < 1$ when $y = 1$.
+-/
+
+section convergent
+
+namespace real
+
+open int
+
+/-!
+### Convergents: definition and API lemmas
+-/
+
+/-- We give a direct recursive definition of the convergents of the continued fraction
+expansion of a real number `ξ`. The main reason for that is that we want to have the
+convergents as rational numbers; the versions
+`(generalized_continued_fraction.of ξ).convergents` and
+`(generalized_continued_fraction.of ξ).convergents'` always give something of the
+same type as `ξ`. We can then also use dot notation `ξ.convergent n`.
+Another minor reason is that this demonstrates that the proof
+of Legendre's theorem does not need anything beyond this definition.
+We provide a proof that this definition agrees with the other one;
+see `real.continued_fraction_convergent_eq_convergent`.
+(Note that we use the fact that `1/0 = 0` here to make it work for rational `ξ`.) -/
+noncomputable def convergent : ℝ → ℕ → ℚ
+| ξ 0 := ⌊ξ⌋
+| ξ (n + 1) := ⌊ξ⌋ + 1 / convergent (1 / fract ξ) n
+
+/-- The zeroth convergent of `ξ` is `⌊ξ⌋`. -/
+@[simp]
+lemma convergent_zero (ξ : ℝ) : ξ.convergent 0 = ⌊ξ⌋ := rfl
+
+/-- The `(n+1)`th convergent of `ξ` is the `n`th convergent of `1/(fract ξ)`. -/
+@[simp]
+lemma convergent_succ (ξ : ℝ) (n : ℕ) :
+  ξ.convergent (n + 1) = ⌊ξ⌋ + 1 / (1 / fract ξ).convergent n :=
+convergent.equations._eqn_2 ξ n
+
+/-- All convergents of `0` are zero. -/
+@[simp]
+lemma convergent_of_zero (n : ℕ) : convergent 0 n = 0 :=
+begin
+  induction n with n ih,
+  { simp only [convergent_zero, floor_zero, cast_zero], },
+  { simp only [ih, convergent_succ, floor_zero, cast_zero, fract_zero, div_zero, add_zero], }
+end
+
+/-- If `ξ` is an integer, all its convergents equal `ξ`. -/
+@[simp]
+lemma convergent_of_int {ξ : ℤ} (n : ℕ) : convergent ξ n = ξ :=
+begin
+  cases n,
+  { simp only [convergent_zero, floor_int_cast], },
+  { simp only [convergent_succ, floor_int_cast, fract_int_cast, div_zero, convergent_of_zero,
+               add_zero], }
+end
+
+/-!
+Our `convergent`s agree with `generalized_continued_fraction.convergents`.
+-/
+
+open generalized_continued_fraction
+
+/-- The `n`th convergent of the `generalized_continued_fraction.of ξ`
+agrees with `ξ.convergent n`. -/
+lemma continued_fraction_convergent_eq_convergent (ξ : ℝ) (n : ℕ) :
```

Comment:
```suggestion
@[simp] lemma continued_fraction_convergent_eq_convergent (ξ : ℝ) (n : ℕ) :
```

Reply:
I'm not sure that it is clear what the preferred spelling should be; it may depend on the context.